share/man/man4/devtree.4
.Dd May 4, 2012
.Dt DEVTREE 4
.Os
.Sh NAME
.Nm devtree
.Nd interface to the kernel's device tree
.Sh SYNOPSIS
.Cd pseudo-device devtree
.Sh DESCRIPTION
.Nm
is an interface to the device tree generated by boot-time
autoconfiguration and (potentially) mutated at run-time as devices are
added to and removed from hot-plug interfaces such as
.Xr usb 4
or
.Xr pcmcia 4 .
.Pp
When a device unit is opened, commands are issued by writing to it and
responses (and potentially asynchronous messages) are obtained by
reading from it.  Calls such as
.Xr select 2
or
.Xr poll 2
may be used to wait for something to be available to read; nonblocking
mode is supported, but asynchronous mode is not.
.Pp
All the manifest constants described below are obtained by including
.Aq Pa dev/pseudo/devtree-intf.h .
.Pp
Each
.Xr write 2
is a separate command; a command may not be split between writes, and
only one command is processed per
.Xr write 2 .
A command consists of a single octet giving the command code; some
commands take an additional octet of data.  The commands are:
.Bl -tag -width indent
.It Dv DTCMD_VERBOSE
This takes one additional octet of data, which is
.Dv DTCMD_ARG_ON
to turn verbosity on or
.Dv DTCMD_ARG_OFF
to turn verbosity off.  (Other second octets produce an
.Dv EIO
error.)  When verbosity is on, console messages designed to help
someone debugging the driver are generated.  (This is not intended for
general-purpose use.)
.It Dv DTCMD_WATCH
This takes one additional octet of data, which is
.Dv DTCMD_ARG_ON
to turn watching on for this unit or
.Dv DTCMD_ARG_OFF
to turn it off.  (Other second octets produce an
.Dv EIO
error.)  When watching is on, the unit receives messages when devices
are attached and detached; see the message descriptions below.
.It Dv DTCMD_OVERRUN
This takes no additional data.  It is part of the buffer-overrun
indication mechanism; see the
.Dv DTMSG_OVERRUN
description, below, for a full description.
.It Dv DTCMD_TREE
This takes no additional data.  It requests a full dump of the device
tree, which is generated as a burst of one or more
.Dv DTMSG_TREE
messages.  See the
.Dv DTMSG_TREE
description, below, for details (but see also
.Dv DTMSG_OVERRUN Ap s
description for an important note).
.El
.Pp
Messages read from the device take the same basic form, but often have
multiple additional data octets.  The possible message type octets and
their semantics are:
.Bl -tag -width indent
.It Dv DTMSG_ATTACH_REAL
This indicates that a device has attached.  This is a
.Sq real
device, as opposed to a pseudo-device.  There are always at least two
octets of additional data, with a variable number of additional octets
following.  The first data octet gives the length of the device name
string; the string itself appears following the second data octet.  If
the device is a device tree root (ie, it has no parent), the second
data octet is 0xff; otherwise, it is the length of the device's
parent's device name string, with the string's contents following the
device name string's contents.  Except for the case of a parentless
device, the total number of data octets is 2 plus the sum of the first
two data octets (for a parentless device, it's 2 plus the first data
octet).
.It Dv DTMSG_ATTACH_PSEUDO
This indicates that a pseudo-device has attached.  There is always at
least one octet of additional data, with a variable number following.
The first data octet gives the length of the device name string, with
the string itself following immediately.  The number of data octets is
one more than the first data octet.
.It Dv DTMSG_DETACH
This indicates that a device has detached.  The following data is
formatted as for
.Dv DTMSG_ATTACH_PSEUDO ,
giving the name of the device in question.  (When detaching, there is
no explicit indication whether the device is real or pseudo.)
.It Dv DTMSG_OVERRUN
This indicates that the kernel's buffer has overflowed.  No data octets
follow.  Once this is generated, no further messages are generated
until (1) all queued messages are read and (2) a
.Dv DTCMD_OVERRUN
command is written.  (If a
.Dv DTCMD_OVERRUN
command is attempted when something is queued to be read, the write
will fail showing
.Er EINVAL . )
Once a
.Dv DTMSG_OVERRUN
message is generated and before it is cleared with a successful
.Dv DTCMD_OVERRUN ,
no attach/detach messages will be generated even if watching is
enabled, and
.Dv DTCMD_TREE
commands will elicit no
.Dv DTMSG_TREE
messages.
.It Dv DTMSG_TREE
This is used to carry the response to
.Dv DTCMD_TREE .
When
.Dv DTCMD_TREE
is issued, a burst of
.Dv DTMSG_TREE
messages is generated, describing the then-current state of the device
tree.  Each
.Dv DTMSG_TREE
message has at least one octet of data following the
.Dv DTMSG_TREE
type octet.  Except for the last one (see below), each one describes
one device (node in the device tree) and has at least two data octets.
The first data octet gives the length of the device's name; the name
itself consists of that many octets following the second data octet.
If the device has a parent, the second data octet gives the length of
the parent's name, with the name itself following the device name; if
the device has no parent, the second data octet is 0xff, and no parent
name string follows the device's name string.  Once the entire device
tree has been returned, one more
.Dv DTMSG_TREE
message is generated, with only one data octet, that one being 0xff.
.Pp
The correct way to reliably track the state of the device tree is to
first enable watching with
.Dv DTCMD_WATCH ,
then request the device tree with
.Dv DTCMD_TREE .
Then ignore any messages that may appear before the burst of
.Dv DTMSG_TREE
messages; after the final
.Dv DTMSG_TREE ,
track device attachment and detachment via
.Dv DTMSG_ATTACH_REAL ,
.Dv DTMSG_ATTACH_PSEUDO ,
and
.Dv DTMSG_DETACH
messages.  (See the
.Sx BUGS
section for a relevant note.)
.El
.Pp
Normally, there is only one
.Nm
device node in
.Pa /dev
(usually
.Pa /dev/devtree ) ,
since there is usually no need to have more than one process watching
the device tree.  However, the driver does support multiple nodes, up
to a maximum defined in the source (at this writing, 256).
.Sh SEE ALSO
.Xr drvctl 8 ,
which provides complementary data.
.Sh HISTORY
.Nm
appeared in May 2012 in patches to
.Nx 1.4T
and
.Nx 4.0.1 .
.Sh BUGS
When using
.Dv DTCMD_TREE as described above, there is a small risk of overrun.
While the device tree is always returned in its entirety, even if the
amount of data would overrun the usual buffer limit, if it is large
enough to overflow and then a device attach or detach occurs before
userland reads enough to drain the queue below the buffer limit, an
overrun will occur.
.Pp
Should be a cloning device rather than having multiple traditional
units.


sys/dev/pseudo/devtree.c
#include <dev/pseudo/devtree-vers.h>

#define DT_INCLUDE_INCLUDES
#include <dev/pseudo/devtree-vers.h>

#include <dev/pseudo/devtree-intf.h>

#define MAXUNIT 256
#define MAXQUEUE 8192

typedef struct softc SOFTC;
typedef struct dblk DBLK;

struct softc {
  int unit;
  unsigned int flags;
#define SF_OPEN    0x00000001
#define SF_NBIO    0x00000002
#define SF_WATCH   0x00000004
#define SF_OVERRUN 0x00000008
  DBLK *dq;
  DBLK **dqt;
  int dqlen;
  struct selinfo rsel;
  } ;

struct dblk {
  DBLK *link;
  int len;
  int ptr;
  } ;

static SOFTC **dtscv;
static int ndevtree;
static int hookid;

int devtree_verbose = 0;

void devtreeattach(int);

void devtreeattach(int n)
{
 ndevtree = 0;
 dtscv = 0;
 hookid = -1;
}

#ifdef DIAGNOSTIC

#define SC_DBQLEN_CHECK(sc) do { sc_dbqlen_check((sc)); } while (0)

static void sc_dbqlen_check(SOFTC *sc)
{
 int n;
 DBLK *db;

 n = 0;
 for (db=sc->dq;db;db=db->link) n += db->len - db->ptr;
 if (n != sc->dqlen)
  { printf("*** devtree driver found qlen %d should be %d\n",sc->dqlen,n);
    sc->dqlen = n;
  }
}

#else

#define SC_DBQLEN_CHECK(sc) do { } while (0)

#endif

static SOFTC *ensure_softc(int u)
{
 SOFTC *s;

 if ((u < 0) || (u >= MAXUNIT)) panic("ensure_softc: impossible unit %d",u);
 if (u >= ndevtree)
  { SOFTC **newv;
    newv = malloc((u+1)*sizeof(SOFTC *),M_DEVBUF,M_WAITOK);
    if (ndevtree)
     { bcopy(dtscv,newv,ndevtree*sizeof(SOFTC *));
       free(dtscv,M_DEVBUF);
     }
    dtscv = newv;
    while (ndevtree <= u) newv[ndevtree++] = 0;
  }
 s = dtscv[u];
 if (! s)
  { s = malloc(sizeof(SOFTC),M_DEVBUF,M_WAITOK);
    s->unit = u;
    s->flags = 0;
    s->dq = 0;
    s->dqt = &s->dq;
    s->dqlen = 0;
     { /* How the fsck are you supposed to initialize a struct selinfo?
	  This seems to work, but is it right?  Who knows.... */
       static const struct selinfo selinfo_init;
       s->rsel = selinfo_init;
     }
    dtscv[u] = s;
  }
 return(s);
}

static void flush_dq(SOFTC *sc)
{
 DBLK *d;

 while ((d = sc->dq))
  { sc->dq = d->link;
    free(d,M_DEVBUF);
  }
 sc->dqt = &sc->dq;
 sc->dqlen = 0;
}

static void append_dblk(SOFTC *sc, const void *data, int len)
{
 DBLK *b;

 b = malloc(sizeof(DBLK)+len,M_DEVBUF,M_WAITOK);
 bcopy(data,b+1,len);
 b->len = len;
 b->ptr = 0;
 b->link = 0;
 *sc->dqt = b;
 sc->dqt = &b->link;
 WAKEUP_SELECT(sc);
 wakeup(sc);
}

static void set_overrun(SOFTC *sc)
{
 static const char om = DTMSG_OVERRUN;

 append_dblk(sc,&om,1);
 sc->flags |= SF_OVERRUN;
}

static void tell_watchers(const char *msg, int msglen)
{
 int i;
 SOFTC *sc;

 for (i=ndevtree-1;i>=0;i--)
  { sc = dtscv[i];
    if (sc && ((sc->flags & (SF_WATCH|SF_OVERRUN)) == SF_WATCH))
     { if (sc->dqlen > MAXQUEUE)
	{ set_overrun(sc);
	}
       else
	{ append_dblk(sc,msg,msglen);
	}
     }
  }
}

static void dump_current_tree(SOFTC *sc)
{
 char msg[1+2+(2*sizeof(((device_t)0)->dv_xname))];
 device_t d;
 int l;

 if (sc->flags & SF_OVERRUN) return;
 if (sc->dqlen > MAXQUEUE)
  { set_overrun(sc);
    return;
  }
 msg[0] = DTMSG_TREE;
 TAILQ_FOREACH(d,&alldevs,dv_list)
  { msg[1] = strlen(&d->dv_xname[0]);
    bcopy(&d->dv_xname[0],&msg[3],msg[1]);
    if (d->dv_parent)
     { msg[2] = strlen(&d->dv_parent->dv_xname[0]);
       bcopy(&d->dv_parent->dv_xname[0],&msg[3+msg[1]],msg[2]);
       l = 3 + msg[1]+ msg[2];
     }
    else
     { msg[2] = 0xff;
       l = 3 + msg[1];
     }
    append_dblk(sc,&msg[0],l);
  }
 msg[1] = 0xff;
 append_dblk(sc,&msg[0],2);
}

static void devtree_hook(device_t dev, enum config_hook_op op)
{
 char msg[1+2+(2*sizeof(dev->dv_xname))];
 int l;

 switch (op)
  { case CONFIG_HOOK_ATTACH_REAL:
       msg[0] = DTMSG_ATTACH_REAL;
       msg[1] = strlen(&dev->dv_xname[0]);
       bcopy(&dev->dv_xname[0],&msg[3],msg[1]);
       if (dev->dv_parent)
	{ msg[2] = strlen(&dev->dv_parent->dv_xname[0]);
	  bcopy(&dev->dv_parent->dv_xname[0],&msg[3+msg[1]],msg[2]);
	  l = 3 + msg[1] + msg[2];
	}
       else
	{ msg[2] = 0xff;
	  l = 3 + msg[1];
	}
       tell_watchers(&msg[0],l);
       printf("devtree_hook %s ATTACH_REAL\n",&dev->dv_xname[0]);
       break;
    case CONFIG_HOOK_ATTACH_PSEUDO:
       msg[0] = DTMSG_ATTACH_PSEUDO;
       msg[1] = strlen(&dev->dv_xname[0]);
       bcopy(&dev->dv_xname[0],&msg[2],msg[1]);
       tell_watchers(&msg[0],2+msg[1]);
       printf("devtree_hook %s ATTACH_PSEUDO\n",&dev->dv_xname[0]);
       break;
    case CONFIG_HOOK_DETACH:
       msg[0] = DTMSG_DETACH;
       msg[1] = strlen(&dev->dv_xname[0]);
       bcopy(&dev->dv_xname[0],&msg[2],msg[1]);
       tell_watchers(&msg[0],2+msg[1]);
       printf("devtree_hook %s DETACH\n",&dev->dv_xname[0]);
       break;
    default:
       printf("devtree_hook %s ?%d\n",&dev->dv_xname[0],(int)op);
       break;
  }
}

DEVSW_SCLASS int devtreeopen(dev_t dev, int flag, int mode, PROCTYPE p)
{
 unsigned int u;
 SOFTC *sc;

 u = minor(dev);
 if (u >= MAXUNIT) return(ENXIO);
 sc = ensure_softc(u);
 if (! (sc->flags & SF_OPEN))
  { sc->flags = 0;
    sc->dq = 0;
    sc->dqt = &sc->dq;
    sc->dqlen = 0;
  }
 if (hookid < 0) hookid = config_set_attach_detach_hook(&devtree_hook);
 sc->flags |= SF_OPEN;
 if (devtree_verbose) printf("devtreeopen: %d succeeded\n",u);
 return(0);
}

DEVSW_SCLASS int devtreeclose(dev_t dev, int flag, int mode, PROCTYPE p)
{
 unsigned int u;
 SOFTC *sc;
 int i;

 u = minor(dev);
 if (u >= ndevtree) panic("closing devtree #%d",u);
 sc = dtscv[u];
 sc->flags = 0;
 flush_dq(sc);
 WAKEUP_SELECT(sc);
 wakeup(sc);
 for (i=ndevtree-1;i>=0;i--)
  { sc = dtscv[i];
    if (sc->flags & SF_OPEN) break;
  }
 if (i < 0)
  { config_clear_attach_detach_hook(hookid);
    hookid = -1;
  }
 if (devtree_verbose) printf("devtreeclose: %d succeeded\n",u);
 return(0);
}

DEVSW_SCLASS int devtreeread(dev_t dev, struct uio *uio, int ioflag)
{
 unsigned int u;
 SOFTC *sc;
 int e;
 DBLK *db;
 int n;

 u = minor(dev);
 if (u >= ndevtree) panic("reading devtree #%d",u);
 sc = dtscv[u];
 if (devtree_verbose) printf("devtreeread %d: entry\n",u);
 while (! sc->dq)
  { if (sc->flags & SF_NBIO)
     { if (devtree_verbose) printf("devtreeread %d: EWOULDBLOCK\n",u);
       return(EWOULDBLOCK);
     }
    e = tsleep(sc,PZERO|PCATCH,"acread",0);
    if (e)
     { if (devtree_verbose) printf("devtreeread %d: %d (tsleep)\n",u,e);
       return(e);
     }
  }
 while (uio->uio_resid && (db = sc->dq))
  { n = db->len - db->ptr;
    if (n > uio->uio_resid) n = uio->uio_resid;
    e = uiomove(db->ptr+(char *)(db+1),n,uio);
    if (e)
     { if (devtree_verbose) printf("devtreeread %d: %d (uiomove)\n",u,e);
       return(e);
     }
    db->ptr += n;
    sc->dqlen -= n;
    if (db->ptr >= db->len)
     { if (! (sc->dq = db->link)) sc->dqt = &sc->dq;
       free(db,M_DEVBUF);
     }
  }
 SC_DBQLEN_CHECK();
 if (devtree_verbose) printf("devtreeread %d: done\n",u);
 return(0);
}

DEVSW_SCLASS int devtreewrite(dev_t dev, struct uio *uio, int ioflag)
{
 unsigned int u;
 SOFTC *sc;
 int e;
 char cmd;

 u = minor(dev);
 if (u >= ndevtree) panic("writing devtree #%d",u);
 sc = dtscv[u];
 if (devtree_verbose) printf("devtreewrite %d: entry\n",u);
 if (uio->uio_resid < 1)
  { if (devtree_verbose) printf("devtreewrite %d: zero-length uio\n",u);
    return(0);
  }
 e = uiomove(&cmd,1,uio);
 if (e)
  { if (devtree_verbose) printf("devtreewrite %d: %d (uiomove cmd)\n",u,e);
    return(e);
  }
 switch (cmd)
  { default:
       if (devtree_verbose) printf("devtreewrite %d: EIO (bad cmd)\n",u);
       return(EIO);
       break;
    case DTCMD_VERBOSE:
       if (uio->uio_resid < 1)
	{ if (devtree_verbose) printf("devtreewrite %d: EIO (v without arg)\n",u);
	  return(EIO);
	}
       e = uiomove(&cmd,1,uio);
       if (e)
	{ if (devtree_verbose) printf("devtreewrite %d: %d (uiomove v arg)\n",u,e);
	  return(e);
	}
       switch (cmd)
	{ case DTCMD_ARG_OFF:
	     devtree_verbose = 0;
	     break;
	  case DTCMD_ARG_ON:
	     devtree_verbose = 1;
	     break;
	  default:
	     if (devtree_verbose) printf("devtreewrite %d: EIO (bad v arg)\n",u);
	     return(EIO);
	     break;
	}
       break;
    case DTCMD_WATCH:
       if (uio->uio_resid < 1)
	{ if (devtree_verbose) printf("devtreewrite %d: EIO (w without arg)\n",u);
	  return(EIO);
	}
       e = uiomove(&cmd,1,uio);
       if (e)
	{ if (devtree_verbose) printf("devtreewrite %d: %d (uiomove w arg)\n",u,e);
	  return(e);
	}
       switch (cmd)
	{ case DTCMD_ARG_OFF:
	     sc->flags &= ~SF_WATCH;
	     break;
	  case DTCMD_ARG_ON:
	     sc->flags |= SF_WATCH;
	     break;
	  default:
	     if (devtree_verbose) printf("devtreewrite %d: EIO (bad v arg)\n",u);
	     return(EIO);
	     break;
	}
       break;
    case DTCMD_OVERRUN:
       if (sc->dqlen == 0) sc->flags &= ~SF_OVERRUN; else return(EINVAL);
       break;
    case DTCMD_TREE:
       dump_current_tree(sc);
       break;
  }
 if (devtree_verbose) printf("devtreewrite %d: done\n",u);
 return(0);
}

DEVSW_SCLASS int devtreeioctl(dev_t dev, u_long cmd, caddr_t data, int flag, PROCTYPE p)
{
 unsigned int u;
 SOFTC *sc;

 u = minor(dev);
 if (u >= ndevtree) panic("ioctling devtree #%d",u);
 sc = dtscv[u];
 if (devtree_verbose) printf("devtreeioctl %d: entry\n",u);
 switch (cmd)
  { case FIONBIO:
       if (devtree_verbose) printf("devtreeioctl %d: FIONBIO %s\n",u,(*(int *)data)?"on":"off");
       if (*(int *)data) sc->flags |= SF_NBIO; else sc->flags &= ~SF_NBIO;
       break;
    case FIOASYNC:
       /* annoying bug in fcntl(F_SETFL) on devices in some versions:
	  the device must support both FIONBIO and FIOASYNC ioctls,
	  even if the corresponding fcntl bits are never set, or
	  nonblocking mode is gratuitously turned off! */
       if (*(int *)data) return(EOPNOTSUPP);
       break;
    default:
       if (devtree_verbose) printf("devtreeioctl %d: ENOTTY cmd %08lx\n",u,(unsigned long int)cmd);
       return(ENOTTY);
       break;
  }
 if (devtree_verbose) printf("devtreeioctl %d: done\n",u);
 return(0);
}

DEVSW_SCLASS int devtreepoll(dev_t dev, int events, PROCTYPE p)
{
 unsigned int u;
 SOFTC *sc;
 int revents;

 u = minor(dev);
 if (u >= ndevtree) panic("polling devtree #%d",u);
 sc = dtscv[u];
 if (devtree_verbose) printf("devtreeioctl %d: entry\n",u);
 revents = events & (POLLOUT|POLLWRNORM);
 if (sc->dq)
  { revents |= events & (POLLIN|POLLRDNORM);
  }
 else
  { selrecord(p,&sc->rsel);
  }
 return(revents);
}

DEFINE_CDEVSW
